2021 Security edits (#2586)
HTTPArchive/almanac.httparchive.org: src/content/en/2021/security.md

We crawl the analyzed pages in both desktop and mobile mode, but for a lot of the data they give similar results, so unless otherwise noted, stats presented in this chapter refer to the set of mobile pages. For more information on how the data has been collected, refer to the [Methodology](./methodology) page.

## Transport security

Following the recent trend, we see continuous growth in the number of websites adopting HTTPS this year as well. Transport Layer Security is important to allow secure browsing of websites by ensuring that the resources being served to you and the data sent to the website are untampered in the transit. Almost all major browsers now come with a HTTPS-only setting and increasing warnings are shown to users when HTTP is used by a website instead of HTTPS, thus pushing broader adoption forward.

{{ figure_markup(
  caption="The percentage of requests that use HTTPS on mobile.",

>   content="91.9%",

Reviewer: Shouldn't this be 91.1% since we are saying the percentage of requests on mobile?

Author: Good eyes! Fixed.

{{ figure_markup(
  image="security-tls-version-by-site.png",
  caption="TLS versions usage for sites.",
  description="Bar chart showing that on desktop 60.4% of sites use TLSv1.3, while 39.6% use TLSv1.2. On mobile the figures are 62.1% and 37.8% respectively. TLSv1.0, TLSv1.1 barely register though there is a very small amount of QUIC usage (less than 0.1% on desktop and mobile).",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vR38ZfiZXxdGuzG4ywCEPKIU1Wl6E2bZwcQPyALavGq5q948gmWb8sT-Xo5T6K5z8smKPg6EKxV0JUI/pubchart?oid=1868638312&format=interactive",
  sheets_gid="1012588329",
  sql_file="tls_versions_pages.sql"
  )
}}

> 60.4% of pages in desktop and 62.1% of pages in mobile are now using TLSv1.3, making it the majority protocol version over of TLSv1.2. The number of pages using TLSv1.3 has increased approximately 20% since [last year](../2020/security#protocol-versions) when we saw 43.2% and 45.4% respectively.

Reviewer: ```suggestion
60.4% of pages in desktop and 62.1% of pages in mobile are now using TLSv1.3, making it the majority protocol version over TLSv1.2. The number of pages using TLSv1.3 has increased approximately 20% since [last year](../2020/security#protocol-versions) when we saw 43.2% and 45.4% respectively.
```

Without properly securing cookies, an attacker can hijack a session and send unwanted changes to the server by impersonating the user. It can also lead to <a hreflang="en" href="https://owasp.org/www-community/attacks/csrf">_Cross-Site Request Forgery_</a> attacks, whereby the user's browser inadvertently sends a request, including the cookies, unbeknownst to the user.

Several other types of attacks rely on the inclusion of cookies in cross-site requests, such as <a hreflang="en" href="https://www.usenix.org/system/files/conference/usenixsecurity15/sec15-paper-lekies.pdf">_Cross-Site Script Inclusion_</a> (XSSI) and various techniques in the <a hreflang="en" href="https://xsleaks.dev/">_XS-Leaks_</a> vulnerability class.

You can ensure that cookies are sent securely and aren't accessed by unintended parties or scripts by adding certain attributes or prefixes.

{{ figure_markup(
  image="security-httponly-secure-samesite-cookie-usage.png",

>   caption="Cookie attributes (mobile).",

Reviewer: ```suggestion
  caption="Cookie attributes (desktop).",
```

This graph is based on desktop data.

    <tbody>
      <tr>
        <td>www.gstatic.com</td>
        <td class="numeric">44.3%</td>
        <td class="numeric">44.1%</td>
      </tr>
      <tr>
        <td>cdn.shopify.com</td>
        <td class="numeric">23.4%</td>
        <td class="numeric">23.9%</td>
      </tr>
      <tr>
        <td>code.jquery.com</td>
        <td class="numeric">7.5%</td>
        <td class="numeric">7.5%</td>
      </tr>
      <tr>
        <td>cdnjs.cloudflare.com</td>
        <td class="numeric">7.2%</td>
        <td class="numeric">6.9%</td>
      </tr>
      <tr>
        <td>stackpath.bootstrapcdn.com</td>
        <td class="numeric">2.7%</td>
        <td class="numeric">2.7%</td>
      </tr>
      <tr>
        <td>maxcdn.bootstrapcdn.com</td>
        <td class="numeric">2.2%</td>
        <td class="numeric">2.3%</td>
      </tr>
      <tr>
        <td>cdn.jsdelivr.net</td>
        <td class="numeric">2.1%</td>
        <td class="numeric">2.1%</td>
      </tr>
    </tbody>
  </table>
  <figcaption>{{ figure_link(caption="Most common hosts from which SRI-protected scripts are included.", sheets_gid="303199583", sql_file="sri_popular_hosts.sql") }}</figcaption>
</figure>

> {# TODO Authors - I do not agree with below. Google fonts does not include SRI. Neither does Google Analytics. Will we just remove this?
> 
> We see that `gstatic.com` currently leads the chart. This is probably because Google serves all of its static content (fonts, analytics, etc.) from gstatic.com. Usually when developers embed google fonts, the code they copy has the `integrity` attribute included, which leads to high adoption.
> #}
> 

Author: Can you review or happy to remove this?

```suggestion
```

{{ figure_markup(
  image="security-cryptominer-market-share.png",
  caption="Cryptominer market share (mobile).",

>   description="Pie chart showing CoinImp has 84.9% of market share, CoinHive has 9.0%, JSEcoin has 3.1%, Minero.cc has 1.5% and others have approximately 1.5%.",

Reviewer 2: ```suggestion
  description="Pie chart showing CoinImp has 84.9% of market share, Coinhive has 9.0%, JSEcoin has 3.1%, Minero.cc has 1.5% and others have approximately 1.5%.",
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vR38ZfiZXxdGuzG4ywCEPKIU1Wl6E2bZwcQPyALavGq5q948gmWb8sT-Xo5T6K5z8smKPg6EKxV0JUI/pubchart?oid=1688060716&format=interactive",
  sheets_gid="1998142066",
  sql_file="cryptominer_share.sql"
) }}

> We see that [CoinHive](https://en.wikipedia.org/wiki/Coinhive) community has switched to CoinImp. Therefore, CoinImp has clearly become the market leader (84.9% share). Even though Coinhive has been discontinued, we still see some websites that use CoinHive scripts (Desktop: 5.7%, mobile: 9.0%).

Reviewer 2: 1) the Wikipedia page redirects to Monero, so not too sure how useful this link is (in the least, link to https://en.wikipedia.org/wiki/Monero#Mining_malware, but maybe a link won't be necessary, because:)
2) it might be interesting to provide the context that the Coinhive domain is now [owned by Troy Hunt](https://www.troyhunt.com/i-now-own-the-coinhive-domain-heres-how-im-fighting-cryptojacking-and-doing-good-things-with-content-security-policies/), and that he has set it up to display a banner (even becoming [quite aggressive](https://twitter.com/troyhunt/status/1455658641344761860))
3) capitalization nit (also for the chart):
```suggestion
We see that [Coinhive](https://en.wikipedia.org/wiki/Coinhive) community has switched to CoinImp. Therefore, CoinImp has clearly become the market leader (84.9% share). Even though Coinhive has been discontinued, we still see some websites that use Coinhive scripts (Desktop: 5.7%, mobile: 9.0%).
```

Author: Corrected the chart.

Though these names are wrong in Wappalyzer too so probably should correct there?

@SaptakS could you look at the wording? Link to the Troy Hunt post would be good IMHO.

Reviewer: I have asked @nrllh to take a look at this since he wrote this section.

Author: As discussed on Slack, here's the suggested wording:

```suggestion
We see that [Coinhive](https://en.wikipedia.org/wiki/Monero#Mining_malware) has been surpassed by CoinImp as the dominate cryptomining service. One of the main reasons for this <a hreflang="en" href="https://www.zdnet.com/article/coinhive-cryptojacking-service-to-shut-down-in-march-2019/">was that Coinhive was shoutdown in March 2019</a>. Interestingly, the domain is now <a hreflang="en" href="https://www.troyhunt.com/i-now-own-the-coinhive-domain-heres-how-im-fighting-cryptojacking-and-doing-good-things-with-content-security-policies/">owned by Troy Hunt</a> who is now displaying aggressive banners in the website in effort to make those sites still hosting the Coinhive script (Desktop: 5.7%, mobile: 9.0%) aware that they are—often without their knowledge. This reflects both the prevalence of Coinhive scripts even over two years after ceasing to operate, and the risks of hosting third-party resources that can be taken over should that third party cease to operate. With Coinhive's demise, CoinImp has clearly become the market leader (84.9% share).
```

Reviewer: Updated the section with the suggestions. Thanks @tunetheweb and @VictorLeP !

## `security.txt`

<a hreflang="en" href="https://datatracker.ietf.org/doc/html/draft-foudil-securitytxt-12">`security.txt`</a> is a file format for websites to provide a standard for vulnerability reporting. Website providers can provide contact details, PGP key, policy, and other information in this file. White hat hackers can then use this information to conduct security analyses on these websites or report a vulnerability.

{{ figure_markup(
  image="security-usage-of-well-known-security.png",
  caption="Use of `security.txt`.",
  description="Bar chart showing 4.9% of websites in desktop and 4.9% of websites in mobile have `security.txt` endpoint.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vR38ZfiZXxdGuzG4ywCEPKIU1Wl6E2bZwcQPyALavGq5q948gmWb8sT-Xo5T6K5z8smKPg6EKxV0JUI/pubchart?oid=1958147881&format=interactive",
  sheets_gid="399976976",
  sql_file="well-known_security.sql"
  )
}}

> We see that just under 5% of the websites already use this standard—although it is still a very young standard. We hope to see this number increase in future, since having a `security.txt` shows that the website administrators care about security and are willing to work with white hat hackers to improve security of their website.

Author: ```suggestion
We see that just under 5% of the websites already use this standard—even though it is still a very young standard. We hope to see this number increase in future, since having a `security.txt` is an indication that the website administrators care about security and are willing to work with vulnerability reporters to improve security of their website.
```

Suggest tweaking this slightly @SaptakS

> We see that just under 5% of the websites already use this standard—even though it is still a very young standard. We hope to see this number increase in future, since having a `security.txt` is an indication that the website administrators care about security and are willing to work with vulnerability reporters to improve security of their website.

Author: ```suggestion
We see that just under 5% of the websites return a response when asking for the `/.well-known/securityt.txt` URL. However investigating many of these show they are basically 404 pages that are incorrectly returning a 200 status code so usage is likely much lower.
```

Reviewer: This does make sense to me and explains why the percentage of properties are so less.

{{ figure_markup(
  image="security-usage-of-properties-in-well-known-security.png",
  caption="Use of security.txt properties.",
  description="Bar chart showing use of different properties in security.txt. `signed` is 0.4% in desktop and 0.3% in mobile, Canonical is 2.8% in desktop and 2.7% in mobile, Encryption is 2.6% in desktop and 2.1% in mobile, Expires is 0.9% in desktop and 0.7% in mobile, Policy is 6.5% in desktop and 6.4% in mobile.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vR38ZfiZXxdGuzG4ywCEPKIU1Wl6E2bZwcQPyALavGq5q948gmWb8sT-Xo5T6K5z8smKPg6EKxV0JUI/pubchart?oid=644563241&format=interactive",
  sheets_gid="399976976",
  sql_file="well-known_security.sql"
  )
}}

> We see that `Policy` is the most used property in the `security.txt` files. This property includes a link to the vulnerability disclosure policy for the website that helps researchers understand the reporting practices they need to follow. The other most used properties include `Canonical` and `Encryption`. `Canonical` is used to indicate where the `security.txt` file is located. If the URI used to retrieve the `security.txt` file doesn't match the list URIs in the `Canonical` fields, then the contents of the file should not be trusted. `Encryption` provides the security researchers with an encryption key that they can use for encrypted communication.

Author: ```suggestion
We see that `Policy` is the most used property in the `security.txt` files, but even then it's only used in 6.4% of sites with a `security.txt` URL. This property includes a link to the vulnerability disclosure policy for the website that helps researchers understand the reporting practices they need to follow. This is therefore likely a better indicator of the real usage of `security.txt` since most file are expected to have a `Policy` value, meaning likely closer to 0.3% of all sites have a "real" `security.txt` file, rather than the 5% measured above.

Another interesting point is that when we look at just this subset of "real" security.txt URLs, Tumblr makes up 63%-65% of the usage. It looks like this is set by default for these domains to the Tumblr contact details. This is great on one hand to show how a single platform can drive adoption of these new security features, but on the other hand indicates a further reduction in actual site usage.

The other most used properties include `Canonical` and `Encryption`. `Canonical` is used to indicate where the `security.txt` file is located. If the URI used to retrieve the `security.txt` file doesn't match the list URIs in the `Canonical` fields, then the contents of the file should not be trusted. `Encryption` provides the security researchers with an encryption key that they can use for encrypted communication.
```

  sheets_gid="399976976",
  sql_file="well-known_security.sql"
  )
}}

We see that just under 5% of the websites already use this standard—although it is still a very young standard.

And the following figure shows that the policy is the most used property of utilized `security.txt` files:

{{ figure_markup(
  image="security-usage-of-properties-in-well-known-security.png",
  caption="Use of security.txt properties.",
  description="Bar chart showing use of different properties in security.txt. `signed` is 0.4% in desktop and 0.3% in mobile, Canonical is 2.8% in desktop and 2.7% in mobile, Encryption is 2.6% in desktop and 2.1% in mobile, Expires is 0.9% in desktop and 0.7% in mobile, Policy is 6.5% in desktop and 6.4% in mobile.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vR38ZfiZXxdGuzG4ywCEPKIU1Wl6E2bZwcQPyALavGq5q948gmWb8sT-Xo5T6K5z8smKPg6EKxV0JUI/pubchart?oid=644563241&format=interactive",
  sheets_gid="399976976",
  sql_file="well-known_security.sql"
  )
}}

{# TODO Anything to say about this graph? #}

> 

Author: Anything to say here? Graph feels kind of just dumped there with no discussion on it.
